### PR TITLE
issue#66 - storage::File `isDeleted` property

### LINKED ISSUE #66: AG dataset index import mechanism
Following up on the initial import mechanism from AG bucket to Elsa (#43 and #58), there are more things need to be done to completely import AG data to Elsa.

- [x] Mechanism on registering new Flagships. 

  Elsa would need to know each folder/prefix in S3 belongs to which dataset, by this there might be a mechanism of registering a map between prefix and dataset record. This process could also be the [initial creation](https://github.com/umccr/elsa-data/blob/26a597614fadbd9b53cbbe02da576c01f2d916c1/application/backend/src/api/routes/dataset-routes.ts#L83) of the dataset record for the files to be associated with.
 
- [x] Handle file deletion

	Some data might be requested to be deleted from AG bucket, ideally this should reflect on EdgeDb. Current implementation is again printing it to the console, probably this should be deleted and a popup warning shows the list of files to be deleted? Also to think, what happen to current grant access for the deleted file?
 
- [x] Complete the data model

	Some of the entries are trios (_mat / _pat) so these should go into a 'trio' case family group

- [ ] Get the Files to be included in SubmissionBatch objects that correspond to each folder

- [ ] Handle mysterious/incomplete submission.
  
  Some data may have pairs located in a different key-prefix/directory, ideally this would still be a valid a location but the current script would only consider pair from the same location. These skipped file is only printed in the console and need to be change so it could be inserted to EdgeDb appropriately.
  Note: If considering to find pair from different location, there is a chance that filename could be duplicated as part of reanalysis. This would be hard to identify which pair belongs to which. (Ref: https://trello.com/c/NFSr0Gjq)

- [ ] Decide on how/any phenotype information can come across into Elsa Data and where to source it

	Even basic things like "sex" and "age at sequencing" are not available to this Elsa Data import as it runs purely off S3 genotype information. What can/do we want to do here?

- [ ] DatasetPatient is no longer consented with the study. Current AG implementation is removing all related patient datas. 

	Do we also need to remove the patient record from `elsa`?  Perhaps another flag to show if patient is no longer available? A consent system exist in Elsa should this be enough without any other flag/deletion? 
	
	~ To be discussed with AG

### PR DESCRIPTION
**Pull request by williamputraintan**
[UPDATE 2Dec]

- Add `isDeleted` property in storage::File schema
- Deletion file from AG will now mark storage::File `isDeleted` property to `true`
- Refactoring/renaming AGService in their folder
- Datasets page (in [UI](https://user-images.githubusercontent.com/61998484/205179475-1364133f-81ae-416b-b0db-7f1e838fd4ea.png)) will have an option to show available datasets only or all datasets (included deleted files).
- `/api/datasets/` have a query option to include deleted files (`?includeDeletedFile=true`).
- Minor update on the detailed dataset page on [UI](https://user-images.githubusercontent.com/61998484/205189282-3569e4bc-2343-4e42-8a41-b89bbe086156.png). (Need to do more work on the detailed dataset page but might be on a seperate ticket.)

<details>
<summary>ARCHIVED - 24Nov</summary>

- Add `isAvailable` property in storage::File schema
- Deletion file from AG will now mark storage::File `isAvailable` property to `false`
- Refactoring/renaming AGService in their folder
- Datasets page (in UI) will only show available datasets instead of all datasets. 
- API: 
- `datasets/all` - show datasets summary from all file in db
- `datasets/available` - show datasets summary from only available file

</details>

### CODE REVIEW

**Comment by andrewpatto** [comment]
I'm not sure I'd introduce a new API endpoint - I think it would be better to just have a field in the Dataset return values..
(kind of like isAvailable). So isLoaded or something.

Reason being - the 'load' on the server for adding in the available cases is almost zero (its just direct config file -> JSON) - so I think there's very little downside in just always having a /datasets endpoint that always includes all the datasets (and then just flip the boolean based on whether there is a corresponding Dataset Db record or not).

So that means the frontend will always just be calling the single API endpoint - and can do what ever filtering/visuals it wants.

**Comment by andrewpatto** [comment]
Should 'description' be deleted from the Dataset EdgeDb schema - as that is better to be sourced direct from the config?
Are there any other fields that we previously stored in the EdgeDb schema for dataset that in fact could be removed and put into the config only?

**Comment by andrewpatto** [comment]
The renaming to S3Index is good - we should document it at some point - as there is no really well established standard for how to 'set up' files for a dataset. So this approach with manifests etc is as good a defacto standard as any. (and isn't restricted to only AG as you've noted)

File: application/backend/src/business/db/dataset-queries.ts [hunk 1]
```diff
@@ -93,6 +93,140 @@ export const datasetAllSummaryQuery = e.params(
     }))
 );
 
+export const datasetAvailableSummaryQuery = e.params(
```
Comment (andrewpatto) [inline application/backend/src/business/db/dataset-queries.ts]: Sorry I think I've just realised I've misunderstood the endpoints here??

So the /datasets/available is talking about File availability or dataset availability?

Comment (williamputraintan) [reply application/backend/src/business/db/dataset-queries.ts]: File availability. So it only takes into account only for files that are available in db. (Deleted file doesn't count).

Comment (andrewpatto) [reply application/backend/src/business/db/dataset-queries.ts]: By takes into account - you mean in terms of the 'stats' (x artifacts in blah bytes?). That would be the only difference then?

Comment (andrewpatto) [reply application/backend/src/business/db/dataset-queries.ts]: If so, that's fine - but I don't think I would represent those by two different API endpoints (or at least not named as you have). The 'resource' for `/api/datasets` is 'the dataset'. So I interpret ../available or ../all to mean the availability *of the dataset*.

Instead I think what you are proposing is the ability to alter the meaning of the fields of the resource - so I'd suggest maybe adding a ?param to the endpoint. One default behaviour - and then a parameter '?includeAvailableFileStatsOnly' (or something like that)

File: application/backend/src/business/db/dataset-queries.ts [hunk 2]
```diff
@@ -93,6 +93,140 @@ export const datasetAllSummaryQuery = e.params(
...
+  { limit: e.int32, offset: e.int32 },
```
Comment (andrewpatto) [inline application/backend/src/business/db/dataset-queries.ts]: I think you could unify the 'all' and 'available' query..

So lets imagine we pass in a boolean parameter into the query..

availabilityMatch: e.bool

File: application/backend/src/business/db/dataset-queries.ts [hunk 3]
```diff
@@ -93,6 +93,140 @@ export const datasetAllSummaryQuery = e.params(
...
+  (params) =>
+    e.select(e.dataset.Dataset, (ds) => {
+      const availBclArtifact = e.select(e.lab.ArtifactBcl, (ab) => ({
+        ...ab["*"],
+        filter: e.op(
+          e.op(ab.bclFile.isAvailable, "=", true),
```
Comment (andrewpatto) [inline application/backend/src/business/db/dataset-queries.ts]: And then just `e.op(ab.bclFile.isAvaiable, "or", params.availabilityMatch)`

would give you the ability to have one query that does both?

File: application/backend/dbschema/storage.esdl [hunk 4]
```diff
@@ -17,5 +17,12 @@ module storage {
         required property size -> int64;
 
         required property checksums -> array<tuple<type: ChecksumType, value: str>>;
+
+        # This would be useful to identify file that has been deleted by the submitter
+        # We won't be deleting the actual record so it doesn't mess up with any of the linking
+        # Instead a flag to identify if data is still available
+        required property isAvailable -> bool {
```
Comment (andrewpatto) [inline application/backend/dbschema/storage.esdl]: The other minor thing I worry about here is that it implies an availability status that we in fact can't guarantee.. i.e the file may in fact *not* be available due to AWS S3 deletion (say). So just literally the naming of the field is possibly too strong.

What we really want to do is mark that we have 'softDeleted' the File (or something along those lines?).

Thoughts?

Comment (andrewpatto) [reply application/backend/dbschema/storage.esdl]: An isAvailable flag to me implies that we would be regularly getting Elsa to loop through the File records and actually check that the file is still present or something. (which may not be a bad idea but is not what this feature is about)

**Comment by andrewpatto** [comment]
Can you attach some screenshots of the new datasets table UI

**Review by andrewpatto — APPROVED** [review]

File: application/backend/config/datasets.json5 [hunk 5]
```diff
@@ -22,12 +22,13 @@
...
       storageUriPrefix: "s3://umccr-10f-data-dev",
     },
-    {
-      name: "10C",
-      description: "UMCCR 10C",
-      uri: "urn:fdc:umccr.org:2022:dataset/10c",
-      storageLocation: "aws-s3",
-      storageUriPrefix: "s3://umccr-10c-data-dev",
-    },
+    // Comment the following to show variety of test-data that is not inclduded in config file
```
Comment (andrewpatto) [inline application/backend/config/datasets.json5]: Given that 'dataset not in config' might in the future affect things - do we actually want to make up a completely bogus 4th dataset for this example (10M "UMCCR 10M" the "missing" dataset)?

Comment (andrewpatto) [reply application/backend/config/datasets.json5]: The 10C dataset will be an important test dataset for cancer cases (even though it isn't particularly set up yet)

Comment (williamputraintan) [reply application/backend/config/datasets.json5]: Yes, sorry. Will add a new dataset then.

**Comment by williamputraintan** [comment]
> Can you attach some screenshots of the new datasets table UI


Screen Shot 2022-12-02 at 1 54 27 pm
(Hovering on the icon led to the tooltip)



Also had some SS from the PR description above on the `UI` word (might need to mentioned this above):
- https://user-images.githubusercontent.com/61998484/205179475-1364133f-81ae-416b-b0db-7f1e838fd4ea.png
- https://user-images.githubusercontent.com/61998484/205189282-3569e4bc-2343-4e42-8a41-b89bbe086156.png

**Pull request merged**